File: reference/typescript/src/ssrf.ts
// AWSP Sender-side SSRF defense.
//
// SPEC.md section 10 requires Senders to:
//   1. Resolve the URL hostname to one or more IPs BEFORE connecting.
//   2. Reject the URL if any resolved address is private, reserved, link-local,
//      multicast, or loopback.
//   3. Connect by IP, not hostname (defeats DNS rebinding).
//   4. Refuse non-HTTPS schemes by default; allow http only as an explicit
//      escape hatch (test fixtures, internal opt-in).
//   5. Cap response size, redirect-follow, and total request time. Those are
//      transport-layer concerns; this module covers (1)-(4).
//
// This module is dependency-free. IP-range matching uses node:net.isIP() to
// validate the address shape, then a small CIDR matcher walks the spec's
// blocklist.
//
// Usage:
//   const safeUrl = await assertPublicUrl(receiverUrl);
//   await fetch(safeUrl, { redirect: 'error', ... });

import { isIP, isIPv4, isIPv6 } from 'node:net';

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export type SsrfBlockedReason =
  | 'private_ip'
  | 'invalid_url'
  | 'dns_failure'
  | 'scheme_not_allowed';

export class SsrfBlockedError extends Error {
  readonly reason: SsrfBlockedReason;
  readonly url: string;
  readonly resolvedIp?: string;
  constructor(reason: SsrfBlockedReason, url: string, message: string, resolvedIp?: string) {
    super(message);
    this.name = 'SsrfBlockedError';
    this.reason = reason;
    this.url = url;
    if (resolvedIp !== undefined) this.resolvedIp = resolvedIp;
  }
}

export interface AssertPublicUrlOptions {
  /** Allow http:// (default false). For test fixtures only. */
  allowHttp?: boolean;
  /** Inject a resolver for testing. Default: node:dns/promises lookup. */
  resolve?: (hostname: string) => Promise<string[]>;
}

/**
 * Validate that a URL is safe for outbound webhook delivery per SPEC.md
 * section 10. Returns a URL with the hostname rewritten to the resolved
 * public IP (so the caller's HTTP client connects to that IP, defeating
 * DNS rebinding). Throws SsrfBlockedError on any failure.
 *
 * The returned URL preserves the original hostname in its `host` only if
 * the original input was an IP literal. For DNS-resolved hostnames, the
 * caller MUST forward the original hostname in the Host / SNI headers
 * (e.g. `fetch(safeUrl, { headers: { Host: originalHostname } })`) -- this
 * function does not handle TLS SNI; that is the HTTP client's job.
 */
export async function assertPublicUrl(
  url: string | URL,
  opts: AssertPublicUrlOptions = {},
): Promise<URL> {
  const allowHttp = opts.allowHttp === true;
  const resolver = opts.resolve ?? defaultResolver;

  const inputStr = typeof url === 'string' ? url : url.toString();
  let parsed: URL;
  try {
    parsed = typeof url === 'string' ? new URL(url) : new URL(url.toString());
  } catch {
    throw new SsrfBlockedError('invalid_url', inputStr, 'URL parse failed');
  }

  // Scheme check.
  if (parsed.protocol === 'https:') {
    // OK.
  } else if (parsed.protocol === 'http:') {
    if (!allowHttp) {
      throw new SsrfBlockedError(
        'scheme_not_allowed',
        inputStr,
        'http:// is not allowed (set allowHttp: true to override)',
      );
    }
  } else {
    throw new SsrfBlockedError(
      'scheme_not_allowed',
      inputStr,
      `scheme ${parsed.protocol} is not allowed (https:// only, http:// with allowHttp)`,
    );
  }

  // Hostname required.
  const hostname = parsed.hostname;
  if (hostname.length === 0) {
    throw new SsrfBlockedError('invalid_url', inputStr, 'URL has no hostname');
  }

  // If the hostname is an IP literal, validate it directly. Strip the
  // bracketed IPv6 form that URL.hostname returns ("[::1]" -> "::1").
  const literalIp = stripBrackets(hostname);
  if (isIP(literalIp) !== 0) {
    if (isPrivateIp(literalIp)) {
      throw new SsrfBlockedError(
        'private_ip',
        inputStr,
        `IP ${literalIp} is in a blocked range`,
        literalIp,
      );
    }
    // Already an IP, no DNS step needed; return as-is (URL already shaped).
    return parsed;
  }

  // DNS resolution.
  let addrs: string[];
  try {
    addrs = await resolver(hostname);
  } catch (err) {
    const msg = err instanceof Error ? err.message : 'DNS lookup failed';
    throw new SsrfBlockedError('dns_failure', inputStr, msg);
  }
  if (!Array.isArray(addrs) || addrs.length === 0) {
    throw new SsrfBlockedError('dns_failure', inputStr, 'no addresses returned');
  }

  // Reject if ANY resolved address is in a blocked range. This is
  // intentionally strict -- if a hostname resolves to one public + one
  // private IP, we refuse the URL rather than picking the public one
  // (defense in depth: a multi-record hostname with a private IP is a
  // strong DNS-rebinding signal).
  for (const addr of addrs) {
    if (isIP(addr) === 0) {
      throw new SsrfBlockedError(
        'dns_failure',
        inputStr,
        `resolver returned non-IP value: ${addr}`,
      );
    }
    if (isPrivateIp(addr)) {
      throw new SsrfBlockedError(
        'private_ip',
        inputStr,
        `${hostname} resolves to ${addr} which is in a blocked range`,
        addr,
      );
    }
  }

  // Pick the first resolved address as the connect target. Rewrite the URL
  // so the caller's HTTP client connects to the IP rather than re-resolving
  // the hostname (DNS-rebinding defense).
  const connectIp = addrs[0] as string;
  const rewritten = new URL(parsed.toString());
  rewritten.hostname = isIPv6(connectIp) ? `[${connectIp}]` : connectIp;
  return rewritten;
}

// ---------------------------------------------------------------------------
// Default resolver (node:dns/promises)
// ---------------------------------------------------------------------------

async function defaultResolver(hostname: string): Promise<string[]> {
  // Lazy import so the module doesn't pull dns at import time (slightly
  // friendlier to bundlers that strip unused deps).
  const dns = await import('node:dns/promises');
  // `all: true` returns every A and AAAA record we get back.
  const results = await dns.lookup(hostname, { all: true, verbatim: true });
  return results.map((r) => r.address);
}

// ---------------------------------------------------------------------------
// IP range matching
// ---------------------------------------------------------------------------

/** Strip the bracketed IPv6 wrapper that URL.hostname returns. */
function stripBrackets(host: string): string {
  if (host.length >= 2 && host.charCodeAt(0) === 0x5b /* [ */ && host.endsWith(']')) {
    return host.slice(1, -1);
  }
  return host;
}

/** Returns true if `ip` is in any blocked range from SPEC.md section 10. */
export function isPrivateIp(ip: string): boolean {
  if (isIPv4(ip)) {
    return isPrivateIpv4(ip);
  }
  if (isIPv6(ip)) {
    return isPrivateIpv6(ip);
  }
  // Anything not parseable as IP is treated as private (fail closed).
  return true;
}

// IPv4 blocklist from SPEC.md section 10.
const IPV4_BLOCKED_CIDRS: ReadonlyArray<readonly [string, number]> = [
  ['0.0.0.0', 8], //         "this network" (RFC 1122)
  ['10.0.0.0', 8], //        private (RFC 1918)
  ['100.64.0.0', 10], //     CGNAT (RFC 6598)
  ['127.0.0.0', 8], //       loopback
  ['169.254.0.0', 16], //    link-local (incl. AWS/GCP metadata 169.254.169.254)
  ['172.16.0.0', 12], //     private (RFC 1918)
  ['192.0.0.0', 24], //      IETF Protocol Assignments (RFC 6890)
  ['192.0.2.0', 24], //      TEST-NET-1 (RFC 5737)
  ['192.168.0.0', 16], //    private (RFC 1918)
  ['198.18.0.0', 15], //     benchmark (RFC 2544)
  ['198.51.100.0', 24], //   TEST-NET-2 (RFC 5737)
  ['203.0.113.0', 24], //    TEST-NET-3 (RFC 5737)
  ['224.0.0.0', 4], //       multicast (RFC 5771)
  ['240.0.0.0', 4], //       reserved future (RFC 1112) -- includes 255.255.255.255
];

function isPrivateIpv4(ip: string): boolean {
  const addr = ipv4ToUint32(ip);
  if (addr === undefined) return true; // malformed -> fail closed
  for (const [base, prefix] of IPV4_BLOCKED_CIDRS) {
    const baseInt = ipv4ToUint32(base);
    if (baseInt === undefined) continue; // unreachable -- base list is curated
    if (matchesCidr32(addr, baseInt, prefix)) return true;
  }
  return false;
}

function ipv4ToUint32(ip: string): number | undefined {
  const parts = ip.split('.');
  if (parts.length !== 4) return undefined;
  let out = 0;
  for (const p of parts) {
    if (p.length === 0 || p.length > 3) return undefined;
    if (!/^[0-9]+$/.test(p)) return undefined;
    const n = Number.parseInt(p, 10);
    if (n < 0 || n > 255) return undefined;
    out = (out * 256 + n) >>> 0; // keep unsigned
  }
  return out;
}

function matchesCidr32(addr: number, base: number, prefix: number): boolean {
  if (prefix === 0) return true;
  // JS shift on 32-bit ints: handle the prefix=32 edge to avoid the shift-by-32
  // becoming shift-by-0.
  const mask = prefix === 32 ? 0xffffffff : (~0 << (32 - prefix)) >>> 0;
  return (addr & mask) === (base & mask);
}

// IPv6 blocklist from SPEC.md section 10.
const IPV6_BLOCKED_CIDRS: ReadonlyArray<readonly [string, number]> = [
  ['::', 128], //          unspecified  ::/128
  ['::1', 128], //         loopback     ::1/128
  // ::ffff:0:0/96 -- IPv4-mapped -- handled by the dedicated branch below.
  ['64:ff9b::', 96], //    NAT64 well-known prefix (RFC 6052)
  ['100::', 64], //        discard prefix (RFC 6666)
  ['2001::', 23], //       IETF Protocol Assignments (RFC 2928)
  ['2001:db8::', 32], //   documentation (RFC 3849)
  ['fc00::', 7], //        unique local (RFC 4193)
  ['fe80::', 10], //       link-local (RFC 4291)
  ['ff00::', 8], //        multicast (RFC 4291)
];

function isPrivateIpv6(ip: string): boolean {
  // Normalize to 8 16-bit groups.
  const groups = expandIpv6(ip);
  if (groups === undefined) return true; // fail closed

  // ::ffff:0:0/96 -- IPv4-mapped IPv6 addresses. Apply IPv4 rules to the
  // embedded v4 octets. Per SPEC.md section 10, the v4 rules apply.
  if (
    groups[0] === 0 &&
    groups[1] === 0 &&
    groups[2] === 0 &&
    groups[3] === 0 &&
    groups[4] === 0 &&
    groups[5] === 0xffff
  ) {
    const g6 = groups[6] ?? 0;
    const g7 = groups[7] ?? 0;
    const v4 = `${(g6 >>> 8) & 0xff}.${g6 & 0xff}.${(g7 >>> 8) & 0xff}.${g7 & 0xff}`;
    // Treat any IPv4-mapped as if it were that v4 (whether private or not).
    return isPrivateIpv4(v4);
  }

  for (const [base, prefix] of IPV6_BLOCKED_CIDRS) {
    const baseGroups = expandIpv6(base);
    if (baseGroups === undefined) continue;
    if (matchesCidr128(groups, baseGroups, prefix)) return true;
  }
  return false;
}

/**
 * Parse an IPv6 string into 8 uint16 groups. Returns undefined on malformed
 * input. Handles `::` shorthand and embedded v4 (e.g. `::ffff:1.2.3.4`).
 */
function expandIpv6(ip: string): number[] | undefined {
  // node:net.isIPv6 validates shape; we still need to do the actual parsing.
  if (!isIPv6(ip)) return undefined;

  // Embedded v4: detect a trailing dotted-quad and convert it to two hex
  // groups so the rest of the parser only deals with colon-separated hex.
  let work = ip;
  if (ip.indexOf('.') !== -1) {
    const lastColon = ip.lastIndexOf(':');
    if (lastColon === -1) return undefined;
    const tail = ip.slice(lastColon + 1);
    const tailInt = ipv4ToUint32(tail);
    if (tailInt === undefined) return undefined;
    const hi = ((tailInt >>> 16) & 0xffff).toString(16);
    const lo = (tailInt & 0xffff).toString(16);
    work = `${ip.slice(0, lastColon)}:${hi}:${lo}`;
  }

  // Split on `::` (RFC 4291 allows at most one).
  const parts = work.split('::');
  if (parts.length > 2) return undefined;

  const parseGroup = (g: string): number | undefined => {
    if (g.length === 0 || g.length > 4) return undefined;
    if (!/^[0-9a-fA-F]+$/.test(g)) return undefined;
    return Number.parseInt(g, 16);
  };

  if (parts.length === 1) {
    // No `::`; must be exactly 8 groups.
    const groups = (parts[0] ?? '').split(':');
    if (groups.length !== 8) return undefined;
    const out: number[] = [];
    for (const g of groups) {
      const n = parseGroup(g);
      if (n === undefined) return undefined;
      out.push(n);
    }
    return out;
  }

  // `::` present; left and right may be empty.
  const leftStr = parts[0] ?? '';
  const rightStr = parts[1] ?? '';
  const left = leftStr === '' ? [] : leftStr.split(':');
  const right = rightStr === '' ? [] : rightStr.split(':');

  const leftGroups: number[] = [];
  for (const g of left) {
    const n = parseGroup(g);
    if (n === undefined) return undefined;
    leftGroups.push(n);
  }
  const rightGroups: number[] = [];
  for (const g of right) {
    const n = parseGroup(g);
    if (n === undefined) return undefined;
    rightGroups.push(n);
  }

  const fillCount = 8 - leftGroups.length - rightGroups.length;
  if (fillCount < 0) return undefined;
  return [...leftGroups, ...zeroFill(fillCount), ...rightGroups];
}

function zeroFill(n: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < n; i++) out.push(0);
  return out;
}

function matchesCidr128(addr: number[], base: number[], prefix: number): boolean {
  if (prefix === 0) return true;
  let remaining = prefix;
  for (let i = 0; i < 8 && remaining > 0; i++) {
    const groupBits = remaining >= 16 ? 16 : remaining;
    const mask = groupBits === 16 ? 0xffff : ((0xffff << (16 - groupBits)) & 0xffff);
    if (((addr[i] ?? 0) & mask) !== ((base[i] ?? 0) & mask)) return false;
    remaining -= groupBits;
  }
  return true;
}

File: reference/go/ssrf.go
// AWSP Sender-side SSRF defense (SPEC.md section 10).
//
// AssertPublicURL resolves the URL's hostname, rejects any address that
// falls in a private, reserved, link-local, multicast, or loopback range,
// and returns a *url.URL with the host rewritten to the resolved public
// IP. Connecting to that returned URL by IP -- not by hostname -- is what
// defeats DNS-rebinding (the resolved IP becomes the connect target).
//
// Senders MUST gate every Receiver-supplied URL through this check before
// opening a connection. See SPEC.md section 10 and THREAT_MODEL.md.
//
// Zero new dependencies: stdlib net/netip only.

package awsp

import (
	"context"
	"errors"
	"fmt"
	"net"
	"net/netip"
	"net/url"
	"strings"
)

// SsrfBlockedError is returned by AssertPublicURL when a URL is rejected.
//
// Reason is one of:
//
//   - "invalid_url"         -- the URL itself failed to parse, or carries
//     an empty host, or has an unparsable port.
//   - "scheme_not_allowed"  -- scheme is not https (and AllowHTTP is false
//     for an http URL, or any other scheme).
//   - "dns_failure"         -- hostname resolution returned no addresses
//     or errored.
//   - "private_ip"          -- at least one resolved address is in a
//     blocked range (see SPEC.md section 10).
//
// URL is the original input string (NOT a sanitized form) so callers can
// log it for triage. ResolvedIP is the address that triggered the
// rejection -- empty when the failure happened before resolution
// (invalid_url, scheme_not_allowed, dns_failure).
type SsrfBlockedError struct {
	Reason     string
	URL        string
	ResolvedIP string
}

// Error implements error.
func (e *SsrfBlockedError) Error() string {
	if e.ResolvedIP != "" {
		return fmt.Sprintf("awsp: SSRF blocked %q: %s (resolved %s)", e.URL, e.Reason, e.ResolvedIP)
	}
	return fmt.Sprintf("awsp: SSRF blocked %q: %s", e.URL, e.Reason)
}

// SSRF rejection-reason sentinels for callers that prefer errors.Is over
// inspecting SsrfBlockedError.Reason directly. AssertPublicURL always
// returns *SsrfBlockedError; these sentinels let callers branch on
// category without a string compare.
var (
	ErrSsrfInvalidURL       = errors.New("invalid_url")
	ErrSsrfSchemeNotAllowed = errors.New("scheme_not_allowed")
	ErrSsrfDNSFailure       = errors.New("dns_failure")
	ErrSsrfPrivateIP        = errors.New("private_ip")
)

// Is supports errors.Is matching against the category sentinels above.
func (e *SsrfBlockedError) Is(target error) bool {
	switch target {
	case ErrSsrfInvalidURL:
		return e.Reason == "invalid_url"
	case ErrSsrfSchemeNotAllowed:
		return e.Reason == "scheme_not_allowed"
	case ErrSsrfDNSFailure:
		return e.Reason == "dns_failure"
	case ErrSsrfPrivateIP:
		return e.Reason == "private_ip"
	}
	return false
}

// AssertPublicURLOptions tunes AssertPublicURL behavior.
//
// Zero-value is valid: AllowHTTP defaults false (https-only), and Resolve
// defaults to net.DefaultResolver.LookupHost.
type AssertPublicURLOptions struct {
	// AllowHTTP, if true, permits http:// in addition to https://.
	// Per SPEC.md section 10, http is allowed only on Sender-internal test
	// fixtures or explicit operator opt-in; default is false.
	AllowHTTP bool

	// Resolve overrides the hostname-resolution step. Useful for tests
	// (stub the resolver) and for callers that want a custom DNS path
	// (e.g. a hardened resolver, a local cache). Returns the list of IP
	// strings the hostname maps to.
	//
	// If nil, net.DefaultResolver.LookupHost is used.
	Resolve func(ctx context.Context, hostname string) ([]string, error)
}

// AssertPublicURL parses rawURL, validates the scheme, resolves the
// hostname, rejects every blocked range from SPEC.md section 10, and
// returns the URL with Host rewritten to the resolved public IP.
//
// The returned URL is intended to be the connect target -- using it
// rewrites the connection from "hostname" to a fixed IP, defeating
// DNS-rebinding attacks where a follow-up resolution would return a
// private address.
//
// Note: the returned URL has Host set to "ip:port" (or "[ip]:port" for
// IPv6). Callers that need to preserve the original Host header for the
// HTTP request itself MUST set Request.Host explicitly to the original
// hostname after substituting the IP into the URL -- otherwise virtual-
// hosted servers will see the wrong Host. The original hostname is
// available on the input rawURL.
//
// On failure returns (nil, *SsrfBlockedError).
func AssertPublicURL(ctx context.Context, rawURL string, opts *AssertPublicURLOptions) (*url.URL, error) {
	if opts == nil {
		opts = &AssertPublicURLOptions{}
	}

	u, err := url.Parse(rawURL)
	if err != nil {
		return nil, &SsrfBlockedError{Reason: "invalid_url", URL: rawURL}
	}
	if u.Host == "" || u.Hostname() == "" {
		return nil, &SsrfBlockedError{Reason: "invalid_url", URL: rawURL}
	}

	// Scheme check. https is always allowed; http only if AllowHTTP.
	scheme := strings.ToLower(u.Scheme)
	switch scheme {
	case "https":
		// allowed
	case "http":
		if !opts.AllowHTTP {
			return nil, &SsrfBlockedError{Reason: "scheme_not_allowed", URL: rawURL}
		}
	default:
		return nil, &SsrfBlockedError{Reason: "scheme_not_allowed", URL: rawURL}
	}

	hostname := u.Hostname()

	// If the hostname is already an IP literal, validate it directly --
	// no DNS lookup needed.
	if literal, err := netip.ParseAddr(strings.Trim(hostname, "[]")); err == nil {
		if isBlockedAddr(literal) {
			return nil, &SsrfBlockedError{
				Reason:     "private_ip",
				URL:        rawURL,
				ResolvedIP: literal.String(),
			}
		}
		return rewriteHost(u, literal), nil
	}

	resolve := opts.Resolve
	if resolve == nil {
		resolve = net.DefaultResolver.LookupHost
	}

	addrs, err := resolve(ctx, hostname)
	if err != nil {
		return nil, &SsrfBlockedError{Reason: "dns_failure", URL: rawURL}
	}
	if len(addrs) == 0 {
		return nil, &SsrfBlockedError{Reason: "dns_failure", URL: rawURL}
	}

	// Reject if ANY resolved address falls in a blocked range. This is
	// stricter than "reject if all are blocked" -- one private answer
	// hidden behind a public one is still an SSRF vector if a future
	// connection picks that one.
	var firstPublic netip.Addr
	for _, a := range addrs {
		// LookupHost returns IP strings without zone or brackets.
		ip, err := netip.ParseAddr(a)
		if err != nil {
			// An unparseable address from the resolver is treated as
			// dns_failure -- we cannot reason about its blockedness.
			return nil, &SsrfBlockedError{Reason: "dns_failure", URL: rawURL, ResolvedIP: a}
		}
		if isBlockedAddr(ip) {
			return nil, &SsrfBlockedError{
				Reason:     "private_ip",
				URL:        rawURL,
				ResolvedIP: ip.String(),
			}
		}
		if !firstPublic.IsValid() {
			firstPublic = ip
		}
	}

	return rewriteHost(u, firstPublic), nil
}

// rewriteHost returns a copy of u with Host replaced by the resolved IP
// (preserving the original port). IPv6 addresses are bracketed regardless
// of whether a port is present, so the resulting URL parses unambiguously.
func rewriteHost(u *url.URL, ip netip.Addr) *url.URL {
	out := *u
	port := u.Port()
	hostStr := ip.String()
	if ip.Is6() && !ip.Is4In6() {
		hostStr = "[" + hostStr + "]"
	}
	if port != "" {
		out.Host = hostStr + ":" + port
	} else {
		out.Host = hostStr
	}
	return &out
}

// isBlockedAddr reports whether addr falls in any range SPEC.md section
// 10 mandates Senders reject. Covers IPv4 and IPv6, including
// IPv4-mapped IPv6 (::ffff:0:0/96 -- the inner IPv4 is checked under
// the IPv4 rules).
func isBlockedAddr(addr netip.Addr) bool {
	// Normalize IPv4-mapped IPv6 (::ffff:a.b.c.d) down to the underlying
	// IPv4 -- per spec, "apply the IPv4 rules". Unmap returns the same
	// addr unchanged for non-mapped values.
	addr = addr.Unmap()

	if addr.Is4() {
		return isBlockedV4(addr)
	}
	return isBlockedV6(addr)
}

// blockedV4 is every CIDR from SPEC.md section 10 IPv4 list.
var blockedV4 = []netip.Prefix{
	netip.MustParsePrefix("0.0.0.0/8"),
	netip.MustParsePrefix("10.0.0.0/8"),
	netip.MustParsePrefix("100.64.0.0/10"),  // RFC 6598 (CGNAT)
	netip.MustParsePrefix("127.0.0.0/8"),    // loopback
	netip.MustParsePrefix("169.254.0.0/16"), // link-local (incl. cloud metadata)
	netip.MustParsePrefix("172.16.0.0/12"),
	netip.MustParsePrefix("192.0.0.0/24"),
	netip.MustParsePrefix("192.0.2.0/24"), // TEST-NET-1
	netip.MustParsePrefix("192.168.0.0/16"),
	netip.MustParsePrefix("198.18.0.0/15"),   // benchmarking
	netip.MustParsePrefix("198.51.100.0/24"), // TEST-NET-2
	netip.MustParsePrefix("203.0.113.0/24"),  // TEST-NET-3
	netip.MustParsePrefix("224.0.0.0/4"),     // multicast
	netip.MustParsePrefix("240.0.0.0/4"),     // reserved (covers 255.255.255.255/32)
}

func isBlockedV4(addr netip.Addr) bool {
	for _, p := range blockedV4 {
		if p.Contains(addr) {
			return true
		}
	}
	return false
}

// blockedV6 is every CIDR from SPEC.md section 10 IPv6 list, EXCLUDING
// ::ffff:0:0/96 (handled via Unmap before this function is called).
var blockedV6 = []netip.Prefix{
	netip.MustParsePrefix("::/128"),        // unspecified
	netip.MustParsePrefix("::1/128"),       // loopback
	netip.MustParsePrefix("64:ff9b::/96"),  // NAT64
	netip.MustParsePrefix("100::/64"),      // discard
	netip.MustParsePrefix("2001::/23"),     // IETF protocol assignments
	netip.MustParsePrefix("2001:db8::/32"), // documentation
	netip.MustParsePrefix("fc00::/7"),      // unique-local
	netip.MustParsePrefix("fe80::/10"),     // link-local
	netip.MustParsePrefix("ff00::/8"),      // multicast
}

func isBlockedV6(addr netip.Addr) bool {
	for _, p := range blockedV6 {
		if p.Contains(addr) {
			return true
		}
	}
	return false
}
